packages/andromeda-math/src/matrix.rs

```diff
@@ -0,0 +1,221 @@
+use cosmwasm_schema::{cw_serde, QueryResponses};
+
+#[andr_instantiate]
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub authorized_operator_addresses: Option<Vec<AndrAddr>>,
+}
+
+#[andr_exec]
+#[cw_serde]
+pub enum ExecuteMsg {
+    StoreMatrix { key: Option<String>, data: Matrix },
+    DeleteMatrix { key: Option<String> },
+}
+
+#[andr_query]
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum QueryMsg {
+    #[returns(GetMatrixResponse)]
+    GetMatrix { key: Option<String> },
+    #[returns(Vec<String>)]
+    AllKeys {},
+    #[returns(Vec<String>)]
+    OwnerKeys { owner: AndrAddr },
+}
+
+#[cw_serde]
+pub struct GetMatrixResponse {
+    pub key: String,
+    pub data: Matrix,
+}
+
+#[cw_serde]
+pub struct Matrix(pub Vec<Vec<i64>>);
+
+impl Matrix {
+    pub fn validate_matrix(&self) -> Result<(), ContractError> {
+        let row_length = self.0.first().map_or(0, |row| row.len());
+        if !self.0.iter().all(|row| row.len() == row_length) {
+            return Err(ContractError::CustomError {
+                msg: "All rows in the matrix must have the same number of columns".to_string(),
+            });
+        }
+        Ok(())
+    }
+
+    pub fn validate_add_sub(&self, other: &Matrix) -> Result<(), ContractError> {
+        if self.0.len() != other.0.len() || self.0[0].len() != other.0[0].len() {
+            return Err(ContractError::CustomError {
+                msg: "Can not add or sub this matrix".to_string(),
+            });
+        }
+        Ok(())
+    }
+
+    pub fn validate_mul(&self, other: &Matrix) -> Result<(), ContractError> {
+        if self.0[0].len() != other.0.len() {
+            return Err(ContractError::CustomError {
+                msg: "Can not multiply this matrix".to_string(),
+            });
+        }
+        Ok(())
+    }
+
+    pub fn add(&self, other: &Matrix) -> Result<Matrix, ContractError> {
+        self.validate_add_sub(other)?;
+
+        let matrix_data = self
+            .0
+            .iter()
+            .zip(&other.0)
+            .map(|(row_a, row_b)| {
+                row_a
+                    .iter()
+                    .zip(row_b)
+                    .map(|(a, b)| a.checked_add(*b).unwrap())
+                    .collect()
+            })
+            .collect();
+
+        Ok(Matrix(matrix_data))
+    }
+
+    pub fn sub(&self, other: &Matrix) -> Result<Matrix, ContractError> {
+        self.validate_add_sub(other)?;
+
+        let matrix_data = self
+            .0
+            .iter()
+            .zip(&other.0)
+            .map(|(row_a, row_b)| {
+                row_a
+                    .iter()
+                    .zip(row_b)
+                    .map(|(a, b)| a.checked_sub(*b).unwrap())
+                    .collect()
+            })
+            .collect();
+
+        Ok(Matrix(matrix_data))
+    }
+
+    #[allow(clippy::needless_range_loop)]
+    pub fn mul(&self, other: &Matrix) -> Result<Matrix, ContractError> {
```

Comment:
Are we using these methods anywhere? Should they be exposed via query messages?

Reply:
We can query saved matrix and then calculate(add, sub, mul) and saved calculated result.